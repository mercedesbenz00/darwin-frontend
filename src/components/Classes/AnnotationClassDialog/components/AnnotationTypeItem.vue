<template>
  <div
    v-if="annotationType"
    v-tooltip="disabled ? {
      content: disabledTooltip,
      placement: 'left',
      offset: 10
    } : undefined"
    class="annotation-type-item"
    :class="{
      'annotation-type-item--disabled': disabled,
      'annotation-type-item--selected': selected
    }"
    @click="onClick"
  >
    <type-icon
      class="annotation-type-item__icon"
      :type="typeName"
      :color="iconColor"
    />
    <span class="annotation-type-item__label">{{ typeNameFormatted }}</span>
    <!--
      The v-if here is because there is no tooltip defined for the "link" type,
      so any team which has links enabled will get errors otherwise.
      https://linear.app/v7labs/issue/V7-985/add-design-for-link-tooltip-in-annotation-class-dialog
    -->
    <annotation-type-item-info
      v-if="hasInfo"
      class="annotation-type-item__info"
      :data="annotationType"
    />
  </div>
</template>

<script lang="ts">
import { Component, Prop, Vue } from 'vue-property-decorator'
import { State } from 'vuex-class'

import TypeIcon from '@/components/Common/AnnotationType/TypeIcon.vue'
import {
  AnnotationTypeName,
  AnnotationTypePayload,
  FeaturePayload,
  RootState
} from '@/store/types'
import { annotationTypeRichData, AnnotationTypeRichPayload, formatTypeName } from '@/utils'

import AnnotationTypeItemInfo from './AnnotationTypeItemInfo/AnnotationTypeItemInfo.vue'

@Component({
  name: 'annotation-type-item',
  components: { AnnotationTypeItemInfo, TypeIcon }
})
export default class AnnotationTypeItem extends Vue {
  @Prop({ required: true })
  typeName!: AnnotationTypeName

  get typeNameFormatted (): string {
    return formatTypeName(this.typeName)
  }

  @Prop({ type: Boolean, default: false })
  disabled!: boolean

  @Prop({ default: null })
  disabledTooltip!: string | null

  @Prop({ type: Boolean, default: false })
  selected!: boolean

  get iconColor (): string {
    return this.$theme.getColor('colorFeatherLight')
  }

  get hasInfo (): boolean {
    const typesWithNoInfo: AnnotationTypeName[] = ['graph', 'link', 'string', 'table']
    return !typesWithNoInfo.includes(this.typeName)
  }

  onClick (): void {
    if (this.disabled) { return }
    this.$emit('click', this.typeName)
  }

  @State((state: RootState) => state.features.list)
  features!: FeaturePayload[] | null

  @State((state: RootState) => state.aclass.types)
  annotationTypes!: AnnotationTypePayload[]

  /**
   * The following object maps annotation type names to the features
   * name that enables them.
   */
  featureEnabledTypes = {
    graph: 'DOCUMENTS',
    link: 'LINK_TOOL',
    string: 'DOCUMENTS',
    mask: 'RASTERS',
    table: 'DOCUMENTS'
  }

  /**
   * Computed property which is used to figure out with rich annotation types should be included
   * when some annotation types are guarded behind a feature flag.
   *
   * The return type is an object with the hidden annotation type name as key,
   * and the rich annotation type payload to be shown as value.
   */
  get annotationType (): AnnotationTypeRichPayload | null {
    const { annotationTypes, features, typeName } = this

    for (const [featureEnabledTypeName, featureName] of Object.entries(this.featureEnabledTypes)) {
      if (typeName !== featureEnabledTypeName) { continue }

      const feature = features?.find(f => f.name === featureName)
      if (!feature?.enabled) { return null }
    }

    const type = annotationTypes.find(t => t.name === typeName)
    const richType = annotationTypeRichData.find(r => r.name === typeName)

    if (!type || !richType) { return null }

    const richSubs = type.subs.map((sub) => {
      const richSub = annotationTypeRichData.find(r => r.name === typeName) || {}
      if (sub && richSub) {
        return { ...sub, ...richSub }
      }
      return null
    }).filter((sub): sub is AnnotationTypeRichPayload => !!sub)

    if (type && richType) {
      return {
        ...type,
        ...richType,
        richSubs
      }
    }

    return null
  }
}
</script>

<style lang="scss" scoped>
.annotation-type-item {
  display: inline-flex;
  align-items: center;
  justify-content: center;
  border-radius: 13px;
  padding: 2px 3px;
  border: 1px solid $colorSecondaryLight3;
  transition: background-color .2s;
  cursor: pointer;
  @include noSelect;

  &:hover:not(.annotation-type-item--disabled),
  &:active:not(.annotation-type-item--disabled) {
    border: 1px solid $colorSecondaryLight2;
    background-color: $colorSecondaryLight2;
    transition: none;
  }
}

.annotation-type-item--selected {
  border: 1px solid $colorSecondaryLight;
  background-color: $colorSecondaryLight1;

  &:hover:not(.annotation-type-item--disabled),
  &:active:not(.annotation-type-item--disabled) {
    border: 1px solid $colorSecondaryLight;
    background-color: $colorSecondaryLight1;
  }
}

.annotation-type-item--disabled {
  .annotation-type-item__icon {
    opacity: .3;
  }
  .annotation-type-item__label {
    opacity: .3;
  }
}

.annotation-type-item__icon {
  @include row--center;
  width: 23px;
  height: 23px;
  margin: 0 8px 0 3px;
  object-fit: contain;
}

.annotation-type-item__label {
  flex: 1;
  @include typography(md, default, normal);
  text-align: left;
  color: $colorSecondaryDark;
  white-space: nowrap;
}
</style>
