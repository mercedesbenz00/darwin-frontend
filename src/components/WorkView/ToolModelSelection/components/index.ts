export { default as ClassPill } from './ClassPill.vue'
export { default as ModelIcon } from './ModelIcon.vue'
