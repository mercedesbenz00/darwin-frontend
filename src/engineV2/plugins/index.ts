import attributesPlugin from './attributes'
import boundingBox from './boundingBox'
import boundingBox3d from './boundingBox3d'
import brush from './brush'
import clicker from './click'
import clipboardPlugin from './clipboard'
import commentatorPlugin from './commentator'
import directionalVector from './directionalVector'
import editPlugin from './edit'
import ellipse from './ellipse'
import fieldPlugin from './field'
import inferencePlugin from './inference'
import instanceIDPlugin from './instanceId'
import keypoint from './keypoint'
import link from './link'
import magnifierGlass from './magnifierGlass'
import mask from './mask'
import measuresPlugin from './measures'
import pixelBrush from './pixelBrush'
import polygon from './polygon'
import polyline from './polyline'
import saveAnnotations from './saveAnnotations'
import selectPlugin from './select'
import skeleton from './skeleton'
import tablePlugin from './table'
import tag from './tag'
import textPlugin from './text'
import v7Look from './v7'
import videoPlayerPlugin from './videoPlayer'
import windowLevelPlugin, { WindowLevelPlugin } from './windowLevel'
import zoomPlugin from './zoom'
import zoomCorrection from './zoomCorrection'
export {
  attributesPlugin,
  boundingBox,
  boundingBox3d,
  brush,
  clicker,
  clipboardPlugin,
  commentatorPlugin,
  directionalVector,
  editPlugin,
  ellipse,
  fieldPlugin,
  inferencePlugin,
  instanceIDPlugin,
  keypoint,
  link,
  magnifierGlass,
  mask,
  measuresPlugin,
  pixelBrush,
  polygon,
  polyline,
  saveAnnotations,
  selectPlugin,
  skeleton,
  tablePlugin,
  tag,
  textPlugin,
  v7Look,
  videoPlayerPlugin,
  windowLevelPlugin,
  WindowLevelPlugin,
  zoomPlugin,
  zoomCorrection
}
