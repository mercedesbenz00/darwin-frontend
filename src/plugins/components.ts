import Vue from 'vue'

import NegativeButton from '@/components/Common/Button/V1/NegativeButton.vue'
import PositiveButton from '@/components/Common/Button/V1/PositiveButton.vue'
import PrimaryButton from '@/components/Common/Button/V1/PrimaryButton.vue'
import SecondaryButton from '@/components/Common/Button/V1/SecondaryButton.vue'
import SecondaryLightButton from '@/components/Common/Button/V1/SecondaryLightButton.vue'

export const installCommonComponents = (_Vue: typeof Vue) => {
  _Vue.component('NegativeButton', NegativeButton)
  _Vue.component('PrimaryButton', PrimaryButton)
  _Vue.component('SecondaryButton', SecondaryButton)
  _Vue.component('SecondaryLightButton', SecondaryLightButton)
  _Vue.component('PositiveButton', PositiveButton)
}
