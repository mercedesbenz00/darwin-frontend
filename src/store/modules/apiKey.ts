import { GetterTree, ActionTree, MutationTree } from 'vuex'

import { RootState, TeamPayload } from '@/store/types'
import { api, parseError, errorMessages, copyAttributes, isErrorResponse } from '@/utils'

import { attachPermissionToModel } from './apiKey/actions/attachPermissionToModel'
import { createForModel } from './apiKey/actions/createForModel'
import { detachPermissionFromModel } from './apiKey/actions/detachPermissionFromModel'
import { ApiKeyPayload, ApiKeyPermission, ApiKeyState } from './apiKey/types'

export const getInitialState = (): ApiKeyState => ({ apiKeys: [] })

const state = getInitialState()
const getters: GetterTree<ApiKeyState, RootState> = {}

type CreateParams = { name: string, team: TeamPayload, permissions: ApiKeyPermission[]}

const actions: ActionTree<ApiKeyState, RootState> = {
  /**
   * Retrive all API keys for a team with the specified id
   */
  async getKeys ({ commit }) {
    let response

    try {
      response = await api.get('api_keys')
    } catch (error: unknown) {
      if (!isErrorResponse(error)) { throw error }
      return parseError(error, errorMessages.API_KEYS_LOAD)
    }

    commit('SET_API_KEYS', response.data)

    return response
  },

  /**
   * Create generic API key for the team and permissions.
   */
  async create ({ commit }, { name, permissions, team }: CreateParams) {
    const path = `teams/${team.id}/api_keys`
    const params = { name, permissions }

    let response

    try {
      response = await api.post(path, params)
    } catch (error: unknown) {
      if (!isErrorResponse(error)) { throw error }
      return parseError(error, errorMessages.API_KEY_CREATE)
    }

    commit('PUSH_API_KEY', response.data)
    return response
  },

  attachPermissionToModel,
  createForModel,
  detachPermissionFromModel,

  /**
   * Update API key
   */
  async updateKey ({ commit }, { apiKey, params }) {
    const path = `api_keys/${apiKey.id}`
    let response

    try {
      response = await api.put(path, params)
    } catch (error: unknown) {
      if (!isErrorResponse(error)) { throw error }
      return parseError(error, errorMessages.API_KEY_UPDATE)
    }

    commit('PUSH_API_KEY', response.data)
    return response
  },

  /**
   * Delete specified API key
   */
  async deleteKey ({ commit }, key: ApiKeyPayload) {
    let response

    try {
      response = await api.remove(`api_keys/${key.id}`, {})
    } catch (error: unknown) {
      if (!isErrorResponse(error)) { throw error }
      return parseError(error, errorMessages.API_KEY_DELETE)
    }

    commit('REMOVE_API_KEY', key)

    return response
  }
}

const mutations: MutationTree<ApiKeyState> = {
  SET_API_KEYS (state, data: ApiKeyPayload[]) {
    state.apiKeys = data
  },

  PUSH_API_KEY (state, data: ApiKeyPayload) {
    const index = state.apiKeys.findIndex(k => k.id === data.id)
    if (index > -1) {
      state.apiKeys.splice(index, 1, data)
    } else {
      state.apiKeys.push(data)
    }
  },

  REMOVE_API_KEY (state, key: ApiKeyPayload) {
    const index = state.apiKeys.findIndex(k => k.id === key.id)
    if (index > -1) {
      state.apiKeys.splice(index, 1)
    }
  },

  RESET_ALL (state) {
    copyAttributes(state, getInitialState())
  }
}

export default {
  namespaced: true,
  state,
  getters,
  actions,
  mutations
}
