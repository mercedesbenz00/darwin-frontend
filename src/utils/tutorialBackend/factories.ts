import {
  AnnotateStagePayload,
  CodeStagePayload,
  CompleteStagePayload,
  DatasetImagePayload,
  DatasetItemPayload,
  DatasetItemStatus,
  DatasetItemType,
  ImagePayload,
  ModelStagePayload,
  ReviewStagePayload,
  StageType,
  WorkflowStagePayload,
  WorkflowStageTemplatePayload,
  WorkflowTemplatePayload
} from '@/store/types'

import TutorialError from './TutorialError'
import { dataset } from './data/dataset'

const itemId = (datasetImageId: number) => datasetImageId * 10
const workflowId = (datasetImageId: number) => datasetImageId * 100

const stageId = (datasetImageId: number, stageNumber: number) =>
  workflowId(datasetImageId) + stageNumber

const initializeDatasetImage = (image: ImagePayload): DatasetImagePayload => ({
  dataset_id: dataset.id,
  dataset_video_id: null,
  id: image.id * 10,
  image,
  seq: image.id,
  set: 1
})

const initializeItem = (datasetImage: DatasetImagePayload): DatasetItemPayload => ({
  archived_reason: null,
  archived: false,
  current_workflow: null,
  current_workflow_id: null,
  dataset_id: datasetImage.dataset_id,
  dataset_image_id: datasetImage.id,
  dataset_image: datasetImage,
  dataset_video_id: null,
  dataset_video: null,
  file_size: 0,
  filename: datasetImage.image.original_filename,
  height: datasetImage.image.height,
  id: itemId(datasetImage.id),
  inserted_at: '2020-06-12T00:00:00',
  labels: [],
  path: '/',
  priority: 0,
  seq: datasetImage.seq,
  set: datasetImage.set,
  status: DatasetItemStatus.annotate,
  updated_at: '2020-06-12T00:00:00',
  width: datasetImage.image.width,
  type: DatasetItemType.image
})

const initializeStage = (
  datasetImage: DatasetImagePayload,
  stageTemplate: WorkflowStageTemplatePayload
): WorkflowStagePayload => {
  const { type } = stageTemplate
  const payload = {
    assignee_id: null,
    completed: false,
    completes_at: null,
    dataset_item_id: itemId(datasetImage.id),
    id: stageId(datasetImage.id, stageTemplate.stage_number),
    metadata: {},
    number: stageTemplate.stage_number,
    skipped_reason: null,
    skipped: false,
    template_metadata: stageTemplate.metadata,
    type: stageTemplate.type,
    workflow_id: workflowId(datasetImage.id),
    workflow_stage_template_id: stageTemplate.id
  }

  if (type === StageType.Annotate) { return payload as AnnotateStagePayload }
  if (type === StageType.Code) { return payload as CodeStagePayload }
  if (type === StageType.Complete) { return payload as CompleteStagePayload }
  if (type === StageType.Model) { return payload as ModelStagePayload }
  if (type === StageType.Review) { return payload as ReviewStagePayload }

  throw new Error('Invalid type given')
}

const initializeWorkflow = (
  datasetImage: DatasetImagePayload,
  workflowTemplate: WorkflowTemplatePayload,
  currentStage: number
): Pick<DatasetItemPayload, 'current_workflow' | 'current_workflow_id'> => {
  const annotateTemplate =
    workflowTemplate.workflow_stage_templates.find(t => t.type === StageType.Annotate)

  const reviewTemplate =
    workflowTemplate.workflow_stage_templates.find(t => t.type === StageType.Review)

  const completeTemplate =
    workflowTemplate.workflow_stage_templates.find(t => t.type === StageType.Complete)

  if (!annotateTemplate || !reviewTemplate || !completeTemplate) {
    throw new TutorialError('Initializing workflow using invalid template ')
  }

  const id = workflowId(datasetImage.id)

  const workflow: DatasetItemPayload['current_workflow'] = {
    dataset_item_id: itemId(datasetImage.id),
    id,
    workflow_template_id: workflowTemplate.id,
    current_stage_number: currentStage,
    current_workflow_stage_template_id: annotateTemplate.id,
    status: DatasetItemStatus.annotate,
    stages: {
      1: [initializeStage(datasetImage, annotateTemplate)],
      2: [initializeStage(datasetImage, reviewTemplate)],
      3: [initializeStage(datasetImage, completeTemplate)]
    }
  }

  return { current_workflow: workflow, current_workflow_id: id }
}

export const createItem = (
  image: ImagePayload,
  workflowTeplate: WorkflowTemplatePayload,
  stageNumber: number = 1
): DatasetItemPayload => {
  const datasetImage = initializeDatasetImage(image)
  return {
    ...initializeItem(datasetImage),
    ...initializeWorkflow(datasetImage, workflowTeplate, stageNumber)
  }
}
