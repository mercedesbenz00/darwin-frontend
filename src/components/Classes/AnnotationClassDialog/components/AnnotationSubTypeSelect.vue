<template>
  <annotation-class-section
    v-if="isRendering"
    title="Subtypes"
    optional
  >
    <div class="annotation-types">
      <annotation-type-item
        v-for="type in annotationSubtypes"
        :key="type"
        class="annotation-type-item"
        disabled-tooltip="Cannot remove existing subtypes"
        :type-name="type"
        :disabled="originalSubTypeSelections[type]"
        :selected="typeSelections[type]"
        @click="updateSubAnnotation(type)"
      />
    </div>
  </annotation-class-section>
</template>

<script lang="ts">
import { Component, Prop, Vue } from 'vue-property-decorator'
import { State } from 'vuex-class'

import Info from '@/components/Common/Info.vue'
import { AnnotationTypeName, AnnotationTypePayload, RootState } from '@/store/types'
import { formatTypeName } from '@/utils'

import AnnotationTypeItem from './AnnotationTypeItem.vue'
import AnnotationClassSection from './Common/AnnotationClassSection.vue'

const ORDERED_SUBS: AnnotationTypeName[] =
  ['attributes', 'text', 'directional_vector', 'instance_id']

@Component({
  name: 'annotation-sub-type-select',
  components: { AnnotationClassSection, AnnotationTypeItem, Info }
})
export default class AnnotationSubTypeSelect extends Vue {
  @Prop({ required: true })
  mainAnnotationType!: AnnotationTypeName | null

  /**
   * Once a subtype has been selected and persisted to backend,
   * it can no longer be removed from a class.
   *
   * This mapping is used to ensure this constraint.
   */
  @Prop({ required: true })
  originalSubTypeSelections!: Partial<Record<AnnotationTypeName, boolean>>

  @Prop({ required: true })
  typeSelections!: Partial<Record<AnnotationTypeName, boolean>>

  @State((state: RootState) => state.aclass.types)
  annotationTypes!: AnnotationTypePayload[]

  get annotationSubtypes (): AnnotationTypeName[] {
    const type =
      this.annotationTypes.find(t => t.name === this.mainAnnotationType && t.granularity === 'main')

    if (!type) {
      throw new Error('Invalid main type given as prop')
    }

    const typeSubs = type.subs.map(s => s.name)
    return ORDERED_SUBS.filter(s => typeSubs.includes(s))
  }

  get isRendering (): boolean {
    return this.annotationSubtypes.length > 0
  }

  formatTypeName = formatTypeName

  updateSubAnnotation (typeName: AnnotationTypeName) {
    this.$emit('update:typeSelections', {
      ...this.typeSelections,
      [typeName]: !this.typeSelections[typeName]
    })
  }
}
</script>

<style lang="scss" scoped>
.annotation-types {
  display: block;
  width: 100%;
}

.annotation-type-item {
  width: calc(50% - 10px);
  margin: 5px 0;

  &:nth-child(even) {
    margin-left: 10px;
  }

  &:nth-child(odd) {
    margin-right: 10px;
  }
}
</style>
