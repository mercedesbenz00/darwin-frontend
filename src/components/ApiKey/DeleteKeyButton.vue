<template>
  <div>
    <negative-button
      size="small"
      @click.stop.prevent="onDelete"
    >
      Delete
    </negative-button>
    <modal
      transition="pop-out"
      width="30%"
      height="275"
      :classes="['confirm-revoke']"
      :click-to-close="false"
      :name="modalName"
    >
      <div class="modal__header">
        <p class="modal__header__title">
          Are you sure you wish to delete this key?
        </p>
        <p class="modal__header__text">
          This action cannot be undone.
        </p>
        <div
          class="modal__header__close"
          @click="cancelDelete"
        >
          &#10005;
        </div>
      </div>
      <div class="modal__footer">
        <secondary-button
          class="modal__footer-left"
          @click.stop="cancelDelete"
        >
          Cancel
        </secondary-button>
        <negative-button
          class="modal__footer-right"
          @click.stop="confirmDelete"
        >
          Delete Key
        </negative-button>
      </div>
    </modal>
  </div>
</template>

<script lang="ts">
import { Component, Vue, Prop } from 'vue-property-decorator'

import { ApiKeyPayload } from '@/store/modules/apiKey/types'

/**
 * Rendered next to an ability. Handles deletion.
 *
 * The original use for this component is to render it in the <key-list-item>
 * default slot, in the generalized <key-management> component, accessible from
 * the settings dialog.
 */
@Component({ name: 'delete-key-button' })
export default class DeleteKeyButton extends Vue {
  @Prop({ required: true, type: Object })
  apiKey!: ApiKeyPayload

  get modalName (): string {
    return `confirm-delete-${this.apiKey.id}`
  }

  onDelete () {
    this.$modal.show(this.modalName)
  }

  cancelDelete () {
    this.$modal.hide(this.modalName)
  }

  confirmDelete () {
    this.delete()
    this.$modal.hide(this.modalName)
  }

  deleting: boolean = false

  async delete () {
    this.deleting = true

    const { error } = await this.$store.dispatch('apiKey/deleteKey', this.apiKey)

    if (error) {
      this.$store.dispatch('toast/warning', { content: error.message })
    }

    this.deleting = false
  }
}
</script>
