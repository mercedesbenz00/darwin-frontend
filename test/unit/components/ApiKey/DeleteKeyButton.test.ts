import { createLocalVue, shallowMount } from '@vue/test-utils'
import flushPromises from 'flush-promises'
import { cloneDeep } from 'lodash'
import VModal from 'vue-js-modal'
import Vuex from 'vuex'

import { buildApiKeyPayload } from 'test/unit/factories'

import DeleteKeyButton from '@/components/ApiKey/DeleteKeyButton.vue'
import { installCommonComponents } from '@/plugins/components'
import apiKey from '@/store/modules/apiKey'

const localVue = createLocalVue()
localVue.use(Vuex)
localVue.use(VModal)
installCommonComponents(localVue)

const newStore = () => {
  const store = new Vuex.Store({
    modules: {
      apiKey: { ...apiKey as any, state: cloneDeep(apiKey.state) }
    }
  })
  jest.spyOn(store, 'dispatch').mockResolvedValue({})
  return store
}

const key = buildApiKeyPayload({
  id: 1,
  prefix: 'abcd',
  name: 'Key 1',
  team_id: 1,
  permissions: [['run_inference', 'model:49bff194-465e-4fff-9077-8f6e6666db3a']]
})

const buttonEventMock = {
  preventDefault: jest.fn(),
  stopPropagation: jest.fn()
}

it('matches snapshot', () => {
  const store = newStore()
  const propsData = { apiKey: key }
  const wrapper = shallowMount(DeleteKeyButton, { localVue, store, propsData })
  expect(wrapper).toMatchSnapshot()
})

it('shows and hides modal on revoke-cancel', async () => {
  const store = newStore()
  const propsData = { apiKey: key }
  const mocks = { $modal: { show: jest.fn(), hide: jest.fn() } }
  const wrapper = shallowMount(DeleteKeyButton, { localVue, mocks, store, propsData })

  await wrapper.find('negative-button-stub').vm.$emit('click', buttonEventMock)
  expect(mocks.$modal.show).toHaveBeenCalled()

  await wrapper.find('.modal__footer-left').vm.$emit('click', buttonEventMock)
  expect(mocks.$modal.hide).toHaveBeenCalled()
})

it('dispatches store action on confirmed revoke', async () => {
  const store = newStore()
  const propsData = { apiKey: key }
  const wrapper = shallowMount(DeleteKeyButton, { localVue, store, propsData })

  await wrapper.find('.modal__footer-right').vm.$emit('click', buttonEventMock)
  expect(store.dispatch).toHaveBeenCalledWith('apiKey/deleteKey', key)
})

it('dispatches toast if revoke fails', async () => {
  const store = newStore()
  const dispatch = store.dispatch as jest.Mock
  dispatch.mockResolvedValue({ error: { message: 'Test error' } })

  const propsData = { apiKey: key }
  const wrapper = shallowMount(DeleteKeyButton, { localVue, store, propsData })

  await wrapper.find('.modal__footer-right').vm.$emit('click', buttonEventMock)
  await flushPromises()

  expect(store.dispatch).toHaveBeenCalledWith('toast/warning', { content: 'Test error' })
})
