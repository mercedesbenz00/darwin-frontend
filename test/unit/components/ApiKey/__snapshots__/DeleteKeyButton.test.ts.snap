// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div>
  <negative-button-stub size="small">
    Delete
  </negative-button-stub>
  <modal-stub name="confirm-delete-1" delay="0" overlaytransition="overlay-fade" transition="pop-out" classes="confirm-revoke" minwidth="0" minheight="0" maxwidth="9007199254740991" maxheight="9007199254740991" width="30%" height="275" pivotx="0.5" pivoty="0.5">
    <div class="modal__header">
      <p class="modal__header__title">
        Are you sure you wish to delete this key?
      </p>
      <p class="modal__header__text">
        This action cannot be undone.
      </p>
      <div class="modal__header__close">
        âœ•
      </div>
    </div>
    <div class="modal__footer">
      <secondary-button-stub class="modal__footer-left">
        Cancel
      </secondary-button-stub>
      <negative-button-stub class="modal__footer-right">
        Delete Key
      </negative-button-stub>
    </div>
  </modal-stub>
</div>
`;
